lib/resty/rediscluster.lua

```diff
@@ -21,13 +21,69 @@ local redis_crc = xmodem.redis_crc
 
 local DEFAULT_SHARED_DICT_NAME = "redis_cluster_slot_locks"
 local DEFAULT_REFRESH_DICT_NAME = "refresh_lock"
-local DEFAULT_MAX_REDIRECTION = 5
-local DEFAULT_MAX_CONNECTION_ATTEMPTS = 3
+local DEFAULT_MAX_REDIRECTION = 2
+local DEFAULT_MAX_CONNECTION_ATTEMPTS = 2
 local DEFAULT_KEEPALIVE_TIMEOUT = 55000
 local DEFAULT_KEEPALIVE_CONS = 1000
 local DEFAULT_CONNECTION_TIMEOUT = 1000
 local DEFAULT_SEND_TIMEOUT = 1000
 local DEFAULT_READ_TIMEOUT = 1000
+local DEFAULT_HEALTH_DICT_NAME = "redis_cluster_health"
+local inspect = require("inspect")
+local err_unhealthy_master = "master node is unhealthy"
+local function health_check_timer(premature)
+    if premature then return end
+
+    local health_dict = ngx.shared[DEFAULT_HEALTH_DICT_NAME]
+    if not health_dict then return end
+
+    local all_keys = health_dict:get_keys()
+    for _, key in ipairs(all_keys) do
+        local ip, port = string.match(key, "^(.-):(%d+)$")
+        if not ip or not port then
+            health_dict:delete(key)
+            goto continue
+        end
+        port = tonumber(port)
+        local ok
+        -- Create a new Redis client for each check
+        local red = redis:new()
+        red:set_timeouts(500, 500, 500)  -- 500ms for connect/send/read
+
+        -- Attempt to connect and send PING
+        local ok, err = red:connect(ip, port)
+        if ok then
+            -- Check if PING succeeds
+            local res, err = red:ping()
+            if res == nil then
+                ok = false
+                err = "PING failed"
+            end
+            red:close()  -- Close connection after check
+        end
+        -- Update health status based on check
+        if ok then
+            health_dict:set(key, 0, 0)  -- Healthy: reset failures, no TTL
+        else
+            local failures = health_dict:get(key) or 0
+            health_dict:set(key, failures + 1, 60)  -- Unhealthy: increment failures with TTL
+        end
+
+        ::continue::
+    end
+end
+
+ngx.timer.every(1, health_check_timer)
```

Comment:
every worker will start this timer, but we only need one.

Reply:
we could run the timer in `privileged agent` process

Reply:
Agreed, but the concept of a privileged agent is unique to APISIX, while this library is designed for general OpenResty environments. We need to find a way for APISIX to control the operation of this timer. Here are my suggestions:
```
--- rediscluster.lua
_M.init()
  ngx.timer.every(1, health_check_timer)
end

--- apisix init.lua
function _M.init_worker()
    if process.type() == "privileged agent" then
        require("rediscluster").init()
    end
end
```